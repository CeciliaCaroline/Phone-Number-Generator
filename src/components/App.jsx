import React, { Component } from "react";
import GeneratorButton from "./GeneratorButton";
import GeneratorInput from "./GeneratorInput";
import Sorting from "./Sorting";
import Counter from "./Counter";
import ClearButton from "./ClearButton";
import DownLoadButton from "./DownLoadButton";
import saveAs from "file-saver";

import Grid from "@material-ui/core/Grid";
import Paper from "@material-ui/core/Paper";
import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableRow from "@material-ui/core/TableRow";
import "./app.css";

class App extends Component {
  state = {
    generateNumber: "",
    randomPhoneNumbers: [],
    sort: "",
    generatedNumbers: [],
    max: null,
    min: null,
    total: null,
    error: null
  };
  randomGenerator = async () => {
    let Generated = new Set();
    let uniqueNumber;
    const { generatedNumbers, generateNumber } = this.state;

    if (generateNumber <= 0 || generateNumber > 10000) {
      this.setState({ error: "Input is above or below limit" });
    } 
      this.setState({ error: "" });

    

    while (Generated.size < generateNumber) {
      let randomNumber =
       Math.floor(Math.random() * 9999999999 + 100000000) + 100000000;

      if (String(randomNumber).length === 9) {
        Generated.add(`0${randomNumber}`);
        uniqueNumber = Array.from(Generated);
        
        this.setState({
          randomPhoneNumbers: uniqueNumber,
          generatedNumbers: generatedNumbers.concat(uniqueNumber)
        });

      }
    }
   
  };

  renderPhoneNumbers = () => {
    const { randomPhoneNumbers} = this.state;
    if (randomPhoneNumbers ) {
      return randomPhoneNumbers.map((UniqueNumber, id) => {
        return (
          <TableRow key={id}>
            <TableCell>{UniqueNumber}</TableCell>
          </TableRow>
        );
      });
    }
  };
  downloadNumbers = () => {
    const { generatedNumbers } = this.state;
    if (generatedNumbers.length > 0) {
      saveAs(
        new Blob(generatedNumbers, { type: "text/plain;charset=utf-8" }),
        "generatedNumbers.csv"
      );
    }
  };

  sortGenerated = () => {
    const { sort, randomPhoneNumbers } = this.state;
    if (randomPhoneNumbers.length > 0) {
      if (sort === "ascending") {
        this.setState({
          randomPhoneNumbers: randomPhoneNumbers.sort(this.sortAscendingNumber)
        });
      } else if (sort === "descending") {
        this.setState({
          randomPhoneNumbers: randomPhoneNumbers.sort(this.sortDescendingNumber)
        });
      }
    } else {
      this.setState({ error: "No phone number have been generated" });
    }
  };

  sortDescendingNumber = (a, b) => {
    if (a > b) {
      return -1;
    }
  };

  sortAscendingNumber = (a, b) => {
    if (a < b) {
      return -1;
    }
  };

  clearNumbers = () => {
    this.setState({ randomPhoneNumbers: [], generateNumber: "" });
  };

  numberInputChange = event => {
    if (event.target.value <= 10000 && event.target.value >0 ) {
      this.setState({ generateNumber: event.target.value, error: "" });
    } else {
      this.setState({ error: "Input is above or below limit", generateNumber: event.target.value });
    }
  };

  setSort = event => {
    event.preventDefault();
    this.setState({ sort: event.target.value }, () => this.sortGenerated());
  };

  render() {
    const {
      generatedNumbers,
      generateNumber,
      error,
      sort,
      randomPhoneNumbers
    } = this.state;
    let max;
    let min;
    let total;
    let count;
    if (generatedNumbers.length > 0) {
      max = Math.max(...generatedNumbers);
      min = Math.min(...generatedNumbers);
      total = generatedNumbers.length;
      count = randomPhoneNumbers.length;
    } else {
      max = 0;
      min = 0;
      total = 0;
      count = 0;
    }

    return (
      <Grid container justify="center">
        <Grid item xs={11}>
          <Paper elevation={24}>
            <div className="container">
              <Grid container justify="center">
                <h3>Random Phone Number Generator</h3>
              </Grid>
              <Grid container justify='space-evenly'>
                <Grid item xs={9}>
                  <GeneratorInput
                    onChange={this.numberInputChange}
                    input={generateNumber}
                  />
                </Grid>

                <Grid item xs={3}>
                  <GeneratorButton
                    length={generateNumber}
                    onClick={this.randomGenerator}
                  />
                </Grid>
              </Grid>

              <Grid container >
                <br />
                <span className="error">{error}</span>
                <br />
              </Grid>
              <Grid container  alignItems="center">
                <Grid item xs={7}>
                  <Sorting sorted={sort} onChange={this.setSort} />
                </Grid>
                <Grid item xs={2}>
                  <DownLoadButton
                    onClick={this.downloadNumbers}
                    randomPhoneNumbers={randomPhoneNumbers}
                  />
                </Grid>
                <Grid item xs={3}>
                  <ClearButton
                    onClick={this.clearNumbers}
                    randomPhoneNumbers={randomPhoneNumbers}
                  />
                </Grid>
              </Grid>
              <Grid container>
                <Grid item xs={12}>
                  <Counter min={min} max={max} total={total} count={count} />
                </Grid>
              </Grid>

              <Grid container justify="flex-start">
                <Table>
                  <TableBody>{this.renderPhoneNumbers()}</TableBody>
                </Table>
              </Grid>
            </div>
          </Paper>
        </Grid>
      </Grid>
    );
  }
}

export default App;
