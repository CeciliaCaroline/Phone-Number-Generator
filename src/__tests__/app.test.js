import React from "react";
import ReactDOM from "react-dom";
import App from "../components/App";
import { render, shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
// import saveAs from "file-saver";

import { configure } from "enzyme";

configure({ adapter: new Adapter() });

describe("App", () => {
  let appInstance;
  let appWrapper;
  beforeEach(() => {
    appWrapper = shallow(<App />);
    appInstance = appWrapper.instance();
  });

  it("it renders the app component", () => {
    const wrapper = render(<App />);
    expect(wrapper.find("div").length).toEqual(21);
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<App />, div);
  });

  it("renders correctly", () => {
    const app = renderer.create(<App />).toJSON();
    expect(app).toMatchSnapshot();
  });

  it("changes number of phone numbers to be generated", () => {
    const generatorInput = appWrapper.find("GeneratorInput");
    const event = {
      preventDefault: () => {},
      target: {
        value: 20
      }
    };
    generatorInput.simulate("change", event);
    expect(appWrapper.state().generateNumber).toBe(20);
  });

  it("input value below limit", () => {
    const event = {
      target: {
        value: -1
      }
    };

    // const inputField = appInstance
    // const generatorInput = appWrapper.find("GeneratorInput");
    appInstance.numberInputChange(event)
    expect(appWrapper.state().generateNumber).toEqual(-1);
    expect(appWrapper.state().error).toBe("Input is above or below limit");



  })

  it("set error when input beyond limit", () => {
    let appWrapper = shallow(<App />);
    appWrapper.state().generateNumber = 15000;

    const event = {
      preventDefault: () => {}
    };
    appInstance.randomGenerator(event);
    appWrapper.setState({ error: "Input is above or below limit" });

    expect(appWrapper.state().generateNumber).toEqual(15000);
    expect(appWrapper.state().error).toBe("Input is above or below limit");
  });

  it("generates random phone numbers", () => {
    appWrapper.state().generateNumber = 1;
    const generatorButton = appWrapper.find("GeneratorButton");
    const event = {
      preventDefault: () => {}
    };
    generatorButton.simulate("click", event);
    appWrapper.state().randomPhoneNumbers= ["0312456789"];
    expect(appWrapper.state().randomPhoneNumbers.length).toBe(1);
  });

  it("clears phone numbers ", () => {
    appWrapper.state().randomPhoneNumbers.length = 10;
    const event = {
      preventDefault: () => {}
    };
    appInstance.clearNumbers(event);
    expect(appWrapper.state().randomPhoneNumbers.length).toBe(0);
    expect(appWrapper.state().generateNumber).toEqual("");
  });

  it("sets sorting value", () => {
    const event = {
      preventDefault: () => {},
      target: {
        value: "ascending"
      }
    };
    appInstance.setSort(event);
    expect(appWrapper.state().sort).toBe(event.target.value);
  });

  it("sorts generated numbers by ascending", () => {


    appWrapper.state().randomPhoneNumbers = ["0312456789", "0123456789", "0342456789"]
    appWrapper.state().sort = "ascending"
    appInstance.sortGenerated()
       appWrapper.setState({
      randomPhoneNumbers: ["0123456789", "0312456789", "0342456789"].sort(appInstance.sortAscendingNumber())
    });
    expect(appWrapper.state().randomPhoneNumbers).toEqual([
      "0123456789",
      "0312456789", "0342456789"
    ]);
  });

  it("sorts generated numbers by descending", () => {
    appWrapper.state().randomPhoneNumbers = ["0312456789", "0123456789", "0342456789"]
    appWrapper.state().sort = "descending"
    appInstance.sortGenerated()
       appWrapper.setState({
      randomPhoneNumbers: ["0123456789", "0312456789", "0342456789"].sort(appInstance.sortDescendingNumber())
    });
    expect(appWrapper.state().randomPhoneNumbers).toEqual([
      "0123456789", "0312456789", "0342456789",
       
    ]);
  });

  // it('testing download function',()=>{
  //   appWrapper.state().randomPhoneNumbers.length = 10;
  //   jest.mock('file-saver', ()=>({saveAs: jest.fn()}))
  //   appInstance.downloadNumbers();
  //   expect(saveAs).toHaveBeenCalledWith()
  //   });

});
