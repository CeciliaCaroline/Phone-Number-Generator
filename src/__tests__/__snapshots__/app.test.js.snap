// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="MuiGrid-container-1 MuiGrid-justify-xs-center-18"
>
  <div
    className="MuiGrid-item-2 MuiGrid-grid-xs-11-40"
  >
    <div
      className="MuiPaper-root-98 MuiPaper-elevation24-124 MuiPaper-rounded-99"
    >
      <div
        className="container"
      >
        <div
          className="MuiGrid-container-1 MuiGrid-justify-xs-center-18"
        >
          <h3>
            Random Phone Number Generator
          </h3>
        </div>
        <div
          className="MuiGrid-container-1 MuiGrid-justify-xs-space-evenly-22"
        >
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-9-38"
          >
            <div
              className="MuiFormControl-root-125 MuiFormControl-fullWidth-128"
            >
              <div
                className="MuiInputBase-root-142 MuiInput-root-129 MuiInput-underline-133 MuiInputBase-fullWidth-151 MuiInput-fullWidth-136 MuiInputBase-formControl-143 MuiInput-formControl-130"
                onClick={[Function]}
              >
                <input
                  aria-invalid={false}
                  className="MuiInputBase-input-152 MuiInput-input-137 MuiInputBase-inputType-155 MuiInput-inputType-140"
                  disabled={false}
                  name="generateNumber"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  placeholder="Enter number between 0 and 10000"
                  required={false}
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-3-32"
          >
            <button
              className="MuiButtonBase-root-185 MuiButtonBase-disabled-186 MuiButton-root-159 MuiButton-text-161 MuiButton-flat-164 MuiButton-disabled-179"
              disabled={true}
              onBlur={[Function]}
              onClick={[Function]}
              onContextMenu={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="-1"
              type="button"
            >
              <span
                className="MuiButton-label-160"
              >
                Generate
              </span>
            </button>
          </div>
        </div>
        <div
          className="MuiGrid-container-1"
        >
          <br />
          <span
            className="error"
          />
          <br />
        </div>
        <div
          className="MuiGrid-container-1 MuiGrid-align-items-xs-center-9"
        >
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-7-36"
          >
            <div
              className="MuiFormControl-root-125"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <label
                className="MuiFormLabel-root-199 MuiInputLabel-root-188 MuiInputLabel-formControl-193 MuiInputLabel-animated-196"
                data-shrink={false}
                htmlFor="sorting"
              >
                Sort by:
              </label>
              <div
                className="MuiInputBase-root-142 MuiInput-root-129 MuiInput-underline-133 MuiInputBase-fullWidth-151 MuiInput-fullWidth-136 MuiInputBase-formControl-143 MuiInput-formControl-130"
                onClick={[Function]}
              >
                <div
                  className="MuiSelect-root-206"
                >
                  <div
                    aria-haspopup="true"
                    aria-pressed="false"
                    className="MuiSelect-select-207 MuiSelect-selectMenu-210 MuiInputBase-input-152 MuiInput-input-137"
                    id="select-sorting"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    role="button"
                    tabIndex={0}
                  >
                    <span
                      dangerouslySetInnerHTML={
                        Object {
                          "__html": "&#8203;",
                        }
                      }
                    />
                  </div>
                  <input
                    id="sorting"
                    name="sorting"
                    type="hidden"
                    value=""
                  />
                  <svg
                    aria-hidden="true"
                    className="MuiSvgIcon-root-213 MuiSelect-icon-212"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-2-31"
          >
            
          </div>
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-3-32"
          >
            
          </div>
        </div>
        <div
          className="MuiGrid-container-1"
        >
          <div
            className="MuiGrid-item-2 MuiGrid-grid-xs-12-41"
          >
            <p
              style={
                Object {
                  "display": "inline-block",
                  "padding": "1%",
                  "width": "auto",
                }
              }
            >
              <b>
                Max: 
              </b>
               
              0
            </p>
            <p
              style={
                Object {
                  "display": "inline-block",
                  "padding": "1%",
                  "width": "auto",
                }
              }
            >
              <b>
                Min: 
              </b>
               
              0
            </p>
            <p
              style={
                Object {
                  "display": "inline-block",
                  "padding": "1%",
                  "width": "auto",
                }
              }
            >
              <b>
                Total: 
              </b>
               
              0
            </p>
          </div>
        </div>
        <div
          className="MuiGrid-container-1"
        >
          <table
            className="MuiTable-root-226"
          >
            <tbody
              className="MuiTableBody-root-227"
            />
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
